File: 13_Bounce/python/bounce.py
"""
BOUNCE

A physics simulation

Ported by Dave LeCompte
"""

from typing import Tuple, List

PAGE_WIDTH = 64


def print_centered(msg: str) -> None:
    spaces = " " * ((PAGE_WIDTH - len(msg)) // 2)
    print(spaces + msg)


def print_header(title: str) -> None:
    print_centered(title)
    print_centered("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print()
    print()
    print()


def print_instructions() -> None:
    print("THIS SIMULATION LETS YOU SPECIFY THE INITIAL VELOCITY")
    print("OF A BALL THROWN STRAIGHT UP, AND THE COEFFICIENT OF")
    print("ELASTICITY OF THE BALL.  PLEASE USE A DECIMAL FRACTION")
    print("COEFFICIENCY (LESS THAN 1).")
    print()
    print("YOU ALSO SPECIFY THE TIME INCREMENT TO BE USED IN")
    print("'STROBING' THE BALL'S FLIGHT (TRY .1 INITIALLY).")
    print()


def get_initial_conditions() -> Tuple[float, float, float]:
    delta_t = float(input("TIME INCREMENT (SEC)? "))
    print()
    v0 = float(input("VELOCITY (FPS)? "))
    print()
    coeff_rest = float(input("COEFFICIENT? "))
    print()

    return delta_t, v0, coeff_rest


def print_at_tab(line: str, tab: int, s: str) -> str:
    line += (" " * (tab - len(line))) + s
    return line


def run_simulation(delta_t: float, v0: float, coeff_rest: float) -> None:
    bounce_time: List[float] = [0] * 20  # time of each bounce

    print("FEET")
    print()

    sim_dur = int(70 / (v0 / (16 * delta_t)))
    for i in range(1, sim_dur + 1):
        bounce_time[i] = v0 * coeff_rest ** (i - 1) / 16

    # Draw the trajectory of the bouncing ball, one slice of height at a time
    h: float = int(-16 * (v0 / 32) ** 2 + v0**2 / 32 + 0.5)
    while h >= 0:
        line = ""
        if int(h) == h:
            line += str(int(h))
        total_time: float = 0
        for i in range(1, sim_dur + 1):
            tm: float = 0
            while tm <= bounce_time[i]:
                total_time += delta_t
                if (
                    abs(h - (0.5 * (-32) * tm**2 + v0 * coeff_rest ** (i - 1) * tm))
                    <= 0.25
                ):
                    line = print_at_tab(line, int(total_time / delta_t), "0")
                tm += delta_t
            tm = bounce_time[i + 1] / 2

            if -16 * tm**2 + v0 * coeff_rest ** (i - 1) * tm < h:
                break
        print(line)
        h = h - 0.5

    print("." * (int((total_time + 1) / delta_t) + 1))
    print
    line = " 0"
    for i in range(1, int(total_time + 0.9995) + 1):
        line = print_at_tab(line, int(i / delta_t), str(i))
    print(line)
    print()
    print(print_at_tab("", int((total_time + 1) / (2 * delta_t) - 2), "SECONDS"))
    print()


def main() -> None:
    print_header("BOUNCE")
    print_instructions()

    while True:
        delta_t, v0, coeff_rest = get_initial_conditions()

        run_simulation(delta_t, v0, coeff_rest)
        break


if __name__ == "__main__":
    main()

File: 13_Bounce/perl/bounce.pl
#!/usr/bin/perl

# Bounce program in Perl
# Translated by Kevin Brannen (kbrannen)

use strict;
use warnings;

print "\n";
print " " x 31,"BOUNCE\n";
print " " x 15, "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n";
print "\n\n\n";

print "THIS SIMULATION LETS YOU SPECIFY THE INITIAL VELOCITY\n";
print "OF A BALL THROWN STRAIGHT UP, AND THE COEFFICIENT OF\n";
print "ELASTICITY OF THE BALL.  PLEASE USE A DECIMAL FRACTION\n";
print "COEFFICIENCY (LESS THAN 1).\n\n";
print "YOU ALSO SPECIFY THE TIME INCREMENT TO BE USED IN\n";
print "'STROBING' THE BALL'S FLIGHT (TRY .1 INITIALLY).\n\n";

# deal with basic's tab() for line positioning
#   line = line string we're starting with
#   pos = position to start writing
#   s = string to write
# returns the resultant string, which might not have been changed
sub line_tab
{
    my ($line, $pos, $s) = @_;
    my $len = length($line);
    # if curser is past position, do nothing
    if ($len <= $pos) { $line .= " " x ($pos - $len) . $s; }
    return $line;
}

while (1)
{
    my @T;          # time slice?
    my $time_inc;   # time increment, probably in fractions of seconds
    my $velocity;   # velocity in feet/sec
    my $coeff_elas; # coeeficent of elasticity
    my $line_pos;   # position on line
    my $S1          # duration in full seconds?

    # get input
    print "TIME INCREMENT (SEC, 0=QUIT): "; chomp($time_inc = <>);
    last if ($time_inc == 0);
    print "VELOCITY (FPS): "; chomp($velocity = <>);
    print "COEFFICIENT: "; chomp($coeff_elas = <>);
    if ($coeff_elas >= 1.0 || $coeff_elas <= 0)
    {
        print "COEFFICIENT MUST BE > 0 AND < 1.0\n\n\n";
        next;
    }

    print "\nFEET\n";
    $S1 = int(70.0 / ($velocity / (16.0 * $time_inc)));
    for my $i (1 .. $S1)
    {
        $T[$i] = $velocity * $coeff_elas ** ($i - 1) / 16.0;
    }

    # draw graph
    for (my $height=int(-16.0 * ($velocity / 32.0) ** 2.0 + $velocity ** 2.0 / 32.0 + .5) ; $height >= 0 ; $height -= .5)
    {
        if (int($height) == $height) { print sprintf("%2d", $height); }
        else                         { print "  "; }
        $line_pos = 0;
        my $curr_line = "";
        for my $i (1 .. $S1)
        {
            my $time;
            for ($time=0 ; $time <= $T[$i] ; $time += $time_inc)
            {
                $line_pos += $time_inc;
                if (abs($height - (.5 * (-32) * $time ** 2.0 + $velocity * $coeff_elas ** ($i - 1) * $time)) <= .25)
                {
                    $curr_line = line_tab($curr_line, ($line_pos / $time_inc), "0");
                }
            }
            $time = ($T[$i + 1] // 0) / 2; # we can reach 1 past the end, use 0 if that happens
            last if (-16.0 * $time ** 2.0 + $velocity * $coeff_elas ** ($i - 1) * $time < $height);
        }
        print "$curr_line\n";
    }

    # draw scale
    print "  .";
    print "." x (int($line_pos + 1) / $time_inc + 1), "\n";
    print "  0";
    my $curr_line = "";
    for my $i (1 .. int($line_pos + .9995))
    {
        $curr_line = line_tab($curr_line, int($i / $time_inc), $i);
    }
    print "$curr_line\n";
    print " " x (int($line_pos + 1) / (2 * $time_inc) - 2), "SECONDS\n\n";
}
